[3.5.2] webpack5 启动报错： Cannot find module 'webpack/lib/container/ModuleFederationPlugin'

<!-- 请不要删除自动生成的 Issue 标签 -->
<!-- 请不要删除自动生成的 Issue 标签 -->
### 相关平台
H5


**浏览器版本: chrome103**
**使用框架: Vue 3**

### 复现步骤
使用 taro build --type h5 --watch 启动

### 期望结果
正常编译启动

依赖配置：
```
"dependencies": {
    "@babel/runtime": "^7.18.9",
    "@nutui/nutui-taro": "^3.2.0",
    "@tarojs/components": "3.5.2",
    "@tarojs/mini-runner": "^3.5.2",
    "@tarojs/plugin-framework-vue3": "3.5.2",
    "@tarojs/plugin-html": "^3.5.2",
    "@tarojs/runtime": "3.5.2",
    "@tarojs/taro": "3.5.2",
    "@tarojs/webpack5-runner": "^3.5.2",
    "pinia": "^2.0.17",
    "snowflake-id": "^1.1.0",
    "taro-plugin-pinia": "^1.0.0",
    "url": "^0.11.0",
    "vue": "^3.2.37"
  },
  "devDependencies": {
    "@babel/core": "^7.18.10",
    "@dcasia/mini-program-tailwind-webpack-plugin": "^1.5.2",
    "@tarojs/cli": "3.5.2",
    "@types/webpack-env": "^1.17.0",
    "@typescript-eslint/eslint-plugin": "^5.32.0",
    "@typescript-eslint/parser": "^5.32.0",
    "@vue/babel-plugin-jsx": "^1.1.1",
    "@vue/compiler-sfc": "^3.2.37",
    "babel-plugin-import": "^1.13.5",
    "babel-preset-taro": "3.5.2",
    "cross-env": "^7.0.3",
    "eslint": "^8.21.0",
    "eslint-config-prettier": "^8.5.0",
    "eslint-config-taro": "3.5.2",
    "eslint-define-config": "^1.5.1",
    "eslint-plugin-import": "^2.26.0",
    "eslint-plugin-jest": "^26.8.0",
    "eslint-plugin-prettier": "^4.2.1",
    "eslint-plugin-vue": "^9.3.0",
    "stylelint": "14.9.1",
    "tailwindcss": "^3.1.8",
    "typescript": "^4.7.4",
    "vue-loader": "^17.0.0"
  }
```
### 实际结果
报错：
```
yarn run v1.22.19
$ cross-env NODE_ENV=development API_ENV=beta taro build --type h5 --watch
👽 Taro v3.5.2

node:internal/modules/cjs/loader:933
  const err = new Error(message);
              ^

Error: Cannot find module 'webpack/lib/container/ModuleFederationPlugin'
Require stack:
- D:\ProgramData\git\repository\leiyan\moti-frontend\moti-storesearch-taro\node_modules\@tarojs\webpack5-prebundle\dist\webpack\TaroModuleFederationPlugin.js
- D:\ProgramData\git\repository\leiyan\moti-frontend\moti-storesearch-taro\node_modules\@tarojs\webpack5-prebundle\dist\prebundle\index.js
- D:\ProgramData\git\repository\leiyan\moti-frontend\moti-storesearch-taro\node_modules\@tarojs\webpack5-prebundle\dist\index.js
- D:\ProgramData\git\repository\leiyan\moti-frontend\moti-storesearch-taro\node_modules\@tarojs\webpack5-runner\dist\index.h5.js
- D:\ProgramData\git\repository\leiyan\moti-frontend\moti-storesearch-taro\node_modules\@tarojs\webpack5-runner\index.js
- D:\ProgramData\git\repository\leiyan\moti-frontend\moti-storesearch-taro\node_modules\@tarojs\helper\dist\npm.js
- D:\ProgramData\git\repository\leiyan\moti-frontend\moti-storesearch-taro\node_modules\@tarojs\helper\dist\index.js
- D:\ProgramData\git\repository\leiyan\moti-frontend\moti-storesearch-taro\node_modules\@tarojs\helper\index.js
- D:\ProgramData\git\repository\leiyan\moti-frontend\moti-storesearch-taro\node_modules\@tarojs\cli\dist\util\index.js
- D:\ProgramData\git\repository\leiyan\moti-frontend\moti-storesearch-taro\node_modules\@tarojs\cli\bin\taro
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (D:\ProgramData\git\repository\leiyan\moti-frontend\moti-storesearch-taro\node_modules\@tarojs\webpack5-prebundle\src\webpack\TaroModuleFederationPlugin.ts:8:1)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Object.newLoader [as .js] (D:\ProgramData\git\repository\leiyan\moti-frontend\moti-storesearch-taro\node_modules\pirates\lib\index.js:141:7)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'D:\\ProgramData\\git\\repository\\leiyan\\moti-frontend\\moti-storesearch-taro\\node_modules\\@tarojs\\webpack5-prebundle\\dist\\webpack\\TaroModuleFederationPlugin.js',
    'D:\\ProgramData\\git\\repository\\leiyan\\moti-frontend\\moti-storesearch-taro\\node_modules\\@tarojs\\webpack5-prebundle\\dist\\prebundle\\index.js',
    'D:\\ProgramData\\git\\repository\\leiyan\\moti-frontend\\moti-storesearch-taro\\node_modules\\@tarojs\\webpack5-prebundle\\dist\\index.js',
    'D:\\ProgramData\\git\\repository\\leiyan\\moti-frontend\\moti-storesearch-taro\\node_modules\\@tarojs\\webpack5-runner\\dist\\index.h5.js',
    'D:\\ProgramData\\git\\repository\\leiyan\\moti-frontend\\moti-storesearch-taro\\node_modules\\@tarojs\\webpack5-runner\\index.js',
    'D:\\ProgramData\\git\\repository\\leiyan\\moti-frontend\\moti-storesearch-taro\\node_modules\\@tarojs\\helper\\dist\\npm.js',
    'D:\\ProgramData\\git\\repository\\leiyan\\moti-frontend\\moti-storesearch-taro\\node_modules\\@tarojs\\helper\\dist\\index.js',
    'D:\\ProgramData\\git\\repository\\leiyan\\moti-frontend\\moti-storesearch-taro\\node_modules\\@tarojs\\helper\\index.js',
    'D:\\ProgramData\\git\\repository\\leiyan\\moti-frontend\\moti-storesearch-taro\\node_modules\\@tarojs\\cli\\dist\\util\\index.js',
    'D:\\ProgramData\\git\\repository\\leiyan\\moti-frontend\\moti-storesearch-taro\\node_modules\\@tarojs\\cli\\bin\\taro'
  ]
}
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

Process finished with exit code 1
```


### 环境信息
```
👽 Taro v3.5.2


  Taro CLI 3.5.2 environment info:
    System:
      OS: Windows 10
    Binaries:
      Node: 16.15.1 - C:\Program Files\nodejs\node.EXE
      Yarn: 1.22.19 - C:\Program Files\nodejs\yarn.CMD
      npm: 8.11.0 - C:\Program Files\nodejs\npm.CMD

```

<!-- generated by taro-issues. 请勿修改或删除此行注释 --><!--labels=T-h5,V-3,F-vue3-->

> 我使用的是react. 3.5.1 =》3.5.2 后我这边是出现： Cannot find module 'webpack'。 然后安装了 'webpack'后，编译时等待非常久卡在编译处，出现： Hot Module Replacement (HMR) is not enabled! React Refresh requires HMR to f

意思是3.5.1是好的吗？

> > 我使用的是react. 3.5.1 =》3.5.2 后我这边是出现： Cannot find module 'webpack'。 然后安装了 'webpack'后，编译时等待非常久卡在编译处，出现： Hot Module Replacement (HMR) is not enabled! React Refresh requires HMR to f
> 
> 意思是3.5.1是好的吗？

是的3.5.1是正确的，我回退到3.5.1后编译正常。

--- a/packages/taro-webpack5-prebundle/src/h5.ts
+++ b/packages/taro-webpack5-prebundle/src/h5.ts
@@ -22,18 +22,16 @@
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */
-import { recursiveMerge } from '@tarojs/helper'
 import fs from 'fs-extra'
 import path from 'path'
 import { performance } from 'perf_hooks'
-import webpack, { Stats } from 'webpack'
+import { Configuration, Stats } from 'webpack'
 import webpackDevServer from 'webpack-dev-server'
 import VirtualModulesPlugin from 'webpack-virtual-modules'
 
 import type { IPrebundle } from './prebundle'
 import BasePrebundle, { IPrebundleConfig } from './prebundle'
 import {
-  createResolve,
   flattenId,
   getMfHash,
   parsePublicPath
@@ -62,7 +60,7 @@ export class H5Prebundle extends BasePrebundle<IH5PrebundleConfig> {
     const mode = process.env.NODE_ENV === 'production' ? 'production' : 'development'
     const devtool = this.config.enableSourceMap && 'hidden-source-map'
     const mainBuildOutput = this.chain.output.entries()
-    const output: Exclude<webpack.Configuration['output'], undefined> = {
+    const output: Exclude<Configuration['output'], undefined> = {
       chunkFilename: this.config.chunkFilename,
       chunkLoadingGlobal: mainBuildOutput.chunkLoadingGlobal,
       globalObject: mainBuildOutput.globalObject,
@@ -88,7 +86,7 @@ export class H5Prebundle extends BasePrebundle<IH5PrebundleConfig> {
 
       this.metadata.runtimeRequirements = new Set<string>()
 
-      const compiler = webpack(recursiveMerge(this.chain.toConfig(), {
+      const compiler = this.getRemoteWebpackCompiler({
         cache: {
           type: 'filesystem',
           cacheDirectory: path.join(this.cacheDir, 'webpack-cache'),
@@ -128,7 +126,7 @@ export class H5Prebundle extends BasePrebundle<IH5PrebundleConfig> {
             }
           )
         ]
-      }, customWebpackConfig))
+      }, customWebpackConfig)
       this.metadata.remoteAssets = await new Promise((resolve, reject) => {
         compiler.run((error: Error, stats: Stats) => {
           compiler.close(err => {
@@ -168,7 +166,6 @@ export class H5Prebundle extends BasePrebundle<IH5PrebundleConfig> {
     this.addPlugin('VirtualModule', VirtualModule)
 
     this.isUseCache = true
-    createResolve(this.appPath, this.chain.toConfig().resolve)
 
     /** 扫描出所有的 node_modules 依赖 */
     const entries: string[] = this.getEntries(this.entryPath)

--- a/packages/taro-webpack5-prebundle/src/mini.ts
+++ b/packages/taro-webpack5-prebundle/src/mini.ts
@@ -22,16 +22,14 @@
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */
-import { recursiveMerge } from '@tarojs/helper'
 import fs from 'fs-extra'
 import path from 'path'
 import { performance } from 'perf_hooks'
-import webpack, { ProvidePlugin, Stats } from 'webpack'
+import { ProvidePlugin, Stats } from 'webpack'
 
 import BasePrebundle, { IPrebundleConfig } from './prebundle'
 import { bundle } from './prebundle/bundle'
 import {
-  createResolve,
   flattenId,
   getBundleHash,
   getMfHash
@@ -141,7 +139,7 @@ export class MiniPrebundle extends BasePrebundle<IMiniPrebundleConfig> {
 
       this.metadata.runtimeRequirements = new Set<string>()
 
-      const compiler = webpack(recursiveMerge(this.chain.toConfig(), {
+      const compiler = this.getRemoteWebpackCompiler({
         cache: {
           type: 'filesystem',
           cacheDirectory: path.join(this.cacheDir, 'webpack-cache'),
@@ -170,7 +168,7 @@ export class MiniPrebundle extends BasePrebundle<IMiniPrebundleConfig> {
           ),
           new ProvidePlugin(provideObject)
         ]
-      }, customWebpackConfig))
+      }, customWebpackConfig)
       this.metadata.remoteAssets = await new Promise((resolve, reject) => {
         compiler.run((error: Error, stats: Stats) => {
           compiler.close(err => {
@@ -199,7 +197,6 @@ export class MiniPrebundle extends BasePrebundle<IMiniPrebundleConfig> {
 
   async run () {
     this.isUseCache = true
-    createResolve(this.appPath, this.chain.toConfig().resolve)
 
     /** 扫描出所有的 node_modules 依赖 */
     /**

--- a/packages/taro-webpack5-prebundle/src/prebundle/index.ts
+++ b/packages/taro-webpack5-prebundle/src/prebundle/index.ts
@@ -1,13 +1,13 @@
 import { Config } from '@swc/core'
-import { chalk, fs, readConfig, resolveMainFilePath, terminalLink } from '@tarojs/helper'
+import { chalk, fs, readConfig, recursiveMerge, resolveMainFilePath, terminalLink } from '@tarojs/helper'
 import { IProjectBaseConfig } from '@tarojs/taro/types/compile'
 import { Message } from 'esbuild'
 import path from 'path'
 import { performance } from 'perf_hooks'
-import { EntryObject } from 'webpack'
+import webpack, { EntryObject } from 'webpack'
 import Chain from 'webpack-chain'
 
-import { commitMeta, getBundleHash, getCacheDir, getMeasure, Metadata, sortDeps } from '../utils'
+import { commitMeta, createResolve, getBundleHash, getCacheDir, getMeasure, Metadata, sortDeps } from '../utils'
 import { CollectedDeps, MF_NAME } from '../utils/constant'
 import TaroModuleFederationPlugin from '../webpack/TaroModuleFederationPlugin'
 import { bundle } from './bundle'
@@ -43,6 +43,7 @@ export default class BasePrebundle<T extends IPrebundleConfig = IPrebundleConfig
 
   deps: CollectedDeps = new Map()
   measure: ReturnType<typeof getMeasure>
+  webpackConfig: webpack.Configuration
 
   constructor (protected config: T, protected option: IPrebundle) {
     if (!option.enable) return
@@ -70,6 +71,9 @@ export default class BasePrebundle<T extends IPrebundleConfig = IPrebundleConfig
         Object.assign(this.preMetadata, fs.readJSONSync(this.metadataPath))
       }
     } catch (e) {}
+
+    this.webpackConfig = this.chain.toConfig()
+    createResolve(this.appPath, this.webpackConfig.resolve)
   }
 
   async run () {
@@ -215,5 +219,15 @@ export default class BasePrebundle<T extends IPrebundleConfig = IPrebundleConfig
         remoteAssets: this.metadata.remoteAssets,
         runtimeRequirements: this.metadata.runtimeRequirements || new Set()
       }])
+  }
+
+  getRemoteWebpackCompiler (standard: webpack.Configuration, custom: webpack.Configuration = {}) {
+    /** NOTE: 删除 Host 应用影响打包 Remote 应用的配置 */
+    const inherit = { ...this.webpackConfig }
+    delete inherit.devServer
+    delete inherit.module
+    delete inherit.plugins
+
+    return webpack(recursiveMerge(inherit, standard, custom))
   }
 }

